lang/src/accounts/option.rs

```diff
@@ -0,0 +1,77 @@
+//! Option<T> type for optional accounts.
+//!
+//! # Example
+//! ```ignore
+//! #[derive(Accounts)]
+//! pub struct Example {
+//!     pub my_acc: Option<Account<'info, MyData>>
+//! }
+//! ```
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use solana_program::account_info::AccountInfo;
+use solana_program::instruction::AccountMeta;
+use solana_program::pubkey::Pubkey;
+
+use crate::{Accounts, AccountsClose, AccountsExit, Result, ToAccountInfos, ToAccountMetas};
+
+impl<'info, T: Accounts<'info>> Accounts<'info> for Option<T> {
+    fn try_accounts(
+        program_id: &Pubkey,
+        accounts: &mut &[AccountInfo<'info>],
+        ix_data: &[u8],
+        bumps: &mut BTreeMap<String, u8>,
+        reallocs: &mut BTreeSet<Pubkey>,
+    ) -> Result<Self> {
+        // There are three cases we have to handle. We don't care if
+        // accounts is empty, so if that's the case we return None. This
+        // allows adding optional accounts at the end of the Accounts
+        // struct without causing a breaking change. This is safe and will error
+        // out if a required account is then added after the optional account and
+        // the accounts aren't passed in.
+        if accounts.is_empty() {
+            return Ok(None);
+        }
+
+        // If there are enough accounts, it will check the program_id and return
+        // None if it matches, popping the first account off the accounts vec.
+        if accounts[0].key == program_id {
+            *accounts = &accounts[1..];
+            Ok(None)
+        } else {
+            // If the program_id doesn't equal the account key, we default to
+            // the try_accounts implementation for the inner type and then wrap that with
+            // Some. This should handle all possible valid cases.
+            T::try_accounts(program_id, accounts, ix_data, bumps, reallocs).map(Some)
+        }
+    }
+}
+
+impl<'info, T: ToAccountInfos<'info>> ToAccountInfos<'info> for Option<T> {
+    fn to_account_infos(&self) -> Vec<AccountInfo<'info>> {
+        self.as_ref()
+            .map_or_else(Vec::new, |account| account.to_account_infos())
+    }
+}
+
+impl<T: ToAccountMetas> ToAccountMetas for Option<T> {
+    fn to_account_metas(&self, is_signer: Option<bool>) -> Vec<AccountMeta> {
+        self.as_ref()
+            .expect("Cannot run `to_account_metas` on None")
```

Comment:
Surely this should have the same `map_or_else` statement as `to_account_infos` if they are to have the same behaviours? 

i.e. it should just return an empty vector. Otherwise it's not possible to call `to_account_metas` on any account struct which contains a null option.

It's possible I don't properly understand what `to_account_metas` is supposed to be used for

Reply:
Ah ok, just noticing that what `to_account_infos` and `to_account_metas` returns here is different from what `to_account_infos` and `to_account_metas` return in the `derive(Accounts)` struct generated methods, e.g. in `__cpi_client_accounts`. It's not a huge deal but just want to make sure they both align with each other in a sensible/consistent way.